TU20Bot/Commands/NameMatch.cs

```diff
@@ -0,0 +1,242 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Discord;
+using Discord.Commands;
+using Discord.WebSocket;
+using TU20Bot.Configuration;
+
+namespace TU20Bot.Commands
+{
+    public class NameMatch : ModuleBase<SocketCommandContext>
+    {
+
+        // The max limit of chars set by discord
+        private const int messageLimit = 2000;
+        private Config _config;
+
+        private const int firstNameIndex = 0;
+        private const int lastNameIndex = 1;
+
+        public delegate void IndexMatch(int index);
+        public delegate void IndexMatchDetailed(int index, string firstName, string lastName);
```

Comment:
> Delegates have some overhead that we aren't using here. Func might be more appropriate.

```csharp
public delegate TResult Func<in T,out TResult>(T arg);
public delegate void IndexMatch(int index);
```
[Source](https://docs.microsoft.com/en-us/dotnet/api/system.func-1?view=netcore-3.1)

This sounds interesting, would you mind elaborating? In C# `Func` is a delegate and a delegate defines the data type for the lambda, which would mean that using a custom delegate and a `Func` within the method are functionally equivalent; the only difference is that a `Func` is already defined within the System namespace, and that a Func requires a return value.

In the case for how these delegates were used, using a `Func` or `Action` would mean defining the argument types (and, in the case of `Func`, the return type) within the `nameNatchAlg` method definition itself, or by defining a constant set of listeners and using those. 

In the former case, this concentrates the definition to a localised area of the file, which is not ideal as it becomes very verbose and code is repeated, which reduces maintainability.

In the latter case, access to the contextual variables is lost (such as the message lists), which doesn't automatically work with the Discord-side of the algorithm design as it was, and it would have made the testing more complex.

Reply:
Hmm, I'm not sure about that.

Delegate defines the `+=` operator and works with events while `Func<>` seems to just act aking to a C function pointer.\

You don't need to define it inside the function method itself if you use `using x = y`. It's how C# aliases types iirc.

Reply:
Would you mind elaborating more? A [Func is a delegate](https://docs.microsoft.com/en-us/dotnet/api/system.func-1?view=netcore-3.1), which means that it would also have the `+=` operator.
`Func<int>` would be defined as: `public delegate TResult Func<out TResult>();`.

In the case of using `using MethodToCall = Func<int, int>;` or `using MethodToCall = Action<int, int>;`:
```csharp
// At the top of the namespace block
using MethodToCall = Action<int, int>; 
// -> delegate void System.Action<in T1, in T2>(T1 arg1, T2 arg2)

// Is functionally equivalent to
delegate void MethodToCall(int index, int otherValue); 
// -> delegate void TU20Bot.Commands.NameMatch.MethodToCall(int index, int otherValue)
```

Both are functionally equivalent, however they are different types.

Reply:
Hmm, that's my bad. However, I'm confident there's a way to get a raw function pointer in C#. Func just might not be it.

Use either one then :)